### PR TITLE
force-control  no good parameter of impedance control

### PR DESCRIPTION
**Pull request by TFLQW**
it also should change the urdf model of the table by adding <stiffness value='100000'/>

### CODE REVIEW

**Comment by lrozo** [comment]
@TFLQW is there any reason to have both examples? I mean, it seems one of them does not work as expected due to the parameter selection? So, is there any reason to have both examples in the pull request?

**Comment by TFLQW** [comment]
@lrozo I have deleted the useless file with bad parameters

File: examples/force_control/Force_control_example.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+        # Pseudo-inverse: \hat{J} = J^T (JJ^T + k^2 I)^{-1}
+        Jp = robot.get_damped_least_squares_inverse(J, damping)
+        # Apply the admittance control
+        Fz_current = FT_sensor.sense()[2]  # record the current Fz
+
+        Fz_error = Fz_current - Fz_desired  # record the current error
+        # dv[2] = dv[2] + 0.00016 * Fz_error + 0.0000008 * (Fz_error - Fz_error_old) / dt # 结果较好的dt=2400
+        # dv[2] = 0.0013 * Fz_error + 0.0000020 * (Fz_error - Fz_error_old) / dt  # 结果较好的dt=2400
+        # dv[2] = 0.00093 * Fz_error + 0.000060 * (Fz_error - Fz_error_old) / dt
```
Comment (lrozo) [inline examples/force_control/Force_control_example.py:141]: Do we need all these comments here? I mean, are they useful for something else?

File: examples/force_control/Force_control_example.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        # dv[2] = 0.0052 * Fz_error
+        # sphere.position[2] = sphere.position[2] + 0.00095 * Fz_error + 0.000060 * (Fz_error - Fz_error_old) / dt
+        if flag == 0:
+            Fz_error_integral = Fz_error + Fz_error_old
+            zzz = sphere.position[2] + 0.000001 * Fz_error + 0.000002 * Fz_error_integral
+            Fz_error_old = Fz_error  # record the current error as the old error
+        elif flag == 1:
+            # xyz 3 direction impedance control
```
Comment (lrozo) [inline examples/force_control/Force_control_example.py:149]: May we remove these comments?

File: examples/force_control/Force_control_example.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+            # K = np.array([[20, 0, 0], [0, 20, 0], [0, 0, 20]])
+            # numerator = np.array([[Fz_error[0], 0, 0], [0, Fz_error[1], 0], [0, 0, Fz_error[2]]]) * np.square(dt) \
+            #             + D * dt * dx[:, 1] + M * (2 * dx[:, 1] - dx[:, 2])
+            # denominator = M + D*dt + K*np.square(dt)
+            # dx_ = numerator * np.linalg.inv(denominator)
+            # dx[:, 2] = dx[:, 1]
+            # dx[:, 1] = dx[:, 0]
+            # dx[:, 0] = np.array([dx_[0, 0], dx_[1, 1], dx_[2, 2]])
+
+            M = 1
```
Comment (lrozo) [inline examples/force_control/Force_control_example.py:161]: May you provide some comments on how you came with the values for M (inertia), D (damping), and K (stiffness)? This info may be useful for future users

File: examples/force_control/Force_control_example.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+            D = 9500
+            K = 500000
+            numerator = Fz_error * np.square(dt) + D * dt * detx[1] + M * (2 * detx[1] - detx[2])
```
Comment (lrozo) [inline examples/force_control/Force_control_example.py:164]: May you provide some comments on which formula you are implementing here? Denominator and numerator are not really good names for variables.

File: examples/force_control/Force_control_example.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+            denominator = M + D*dt + K*np.square(dt)
+            detx_ = numerator / denominator
+            print (detx_)
+            detx[2] = detx[1]
+            detx[1] = detx[0]
+            detx[0] = detx_
+
+            zzz = sphere.position[2] + detx[0]
+
+        sphere.position = np.array([0.46 - r * np.sin(w * t + np.pi / 2), r * np.cos(w * t + np.pi / 2), zzz])
```
Comment (lrozo) [inline examples/force_control/Force_control_example.py:174]: where does the 0.46 value come from?

**Comment by TFLQW** [comment]
@lrozo I have added some necessary comments above these questions that you have asked

File: examples/force_control/Force_control_example.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+import numpy as np
+from itertools import count
+
+from pyrobolearn.simulators import Bullet
+from pyrobolearn.worlds import BasicWorld
+from pyrobolearn.robots import KukaIIWA, Body, sensors
+from pyrobolearn.utils.transformation import *
+
+from simulate_test_ur.plotting_ee_FT import EeFtRealTimePlot
```
Comment (lrozo) [inline examples/force_control/Force_control_example.py:14]: You did not include this python file and therefore we are not able to test the pull request functionality. May you please upload it? 
However, I removed the dependencies on this plotting functionality, and the code runs nicely! So, if you just do the suggested changes, we may merge the pull request :)

File: examples/force_control/Force_control_example.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""Inverse kinematics with the Kuka robot where the goal is to follow a moving sphere.
```
Comment (lrozo) [inline examples/force_control/Force_control_example.py:3]: Please update the description of the code to fit the functionalities coded here. Please be as descriptive as possible.

File: examples/force_control/Force_control_example.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+        # set joint positions
+        q = q[qIdx] + dq * dt
+        robot.set_joint_positions(q, joint_ids=joint_ids)
+        if t > 300:
+            break
+        # step in simulation
+        world.step(sleep_dt=dt)
+
+    for t in count():
```
Comment (lrozo) [inline examples/force_control/Force_control_example.py:66]: Please add a comment on what this loop does.

File: examples/force_control/Force_control_example.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+        if FT_sensor.sense() is not None:
+            if FT_sensor.sense()[2] > Fz_desired:
+                break
+        # step in simulation
+        world.step(sleep_dt=dt)
+    sp_z = []
+    num = []
+    detx = np.array([0.0, 0.0, 0.0])
+    for t in count():
```
Comment (lrozo) [inline examples/force_control/Force_control_example.py:107]: Also here, please add a comment on what the loop does.

File: examples/force_control/Force_control_example.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+        Fz_current = FT_sensor.sense()[2]  # record the current Fz
+
+        Fz_error = Fz_current - Fz_desired  # record the current error
+
+        # set the M\D\K parameters by heuristic method, these parameters may have a good result
+        M = 1
+        D = 9500
+        K = 500000
+        # Refer the formula in this article
+        # [1] SONG, Peng; YU, Yueqing; ZHANG, Xuping. A tutorial survey and comparison of impedance control on robotic manipulation. Robotica, 2019, 37.5: 801-836.
```
Comment (lrozo) [inline examples/force_control/Force_control_example.py:141]: You can move the reference to the code description at the beginning of the file and here you only need to refer to [1]

File: examples/force_control/Force_control_example.py (hunk 5 above)
Comment (lrozo) [reply examples/force_control/Force_control_example.py:154]: It is then better to define a variable `circle_center` that has this value. This is to have a readable code. You may apply the same concept for other constants you have along the code.

File: examples/force_control/Force_control_example.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+        # set joint positions
+        q = q[qIdx] + dq * dt
+        robot.set_joint_positions(q, joint_ids=joint_ids)
+
+        # print(Fz_error, dv[2])
+        if t == 800:
+            break
+        # step in simulation
+        world.step(sleep_dt=dt)
+    plt.plot(num, sp_z)
```
Comment (lrozo) [inline examples/force_control/Force_control_example.py:174]: Add some labels to both horizontal and vertical axes of the plot.

File: examples/force_control/Force_control_example.py [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+    # I set the reference orientation to a constant
+    sphere.orientation = np.array([1, 0, 0, 0])
+
+    FT_sensor = sensors.JointForceTorqueSensor(sim, body_id=robot.id, joint_ids=6)
+    # The plotting handle
+    plot = EeFtRealTimePlot(robot, sensor=FT_sensor, forcex=True, forcey=True, forcez=True,
+                            torquex=True, torquey=True, torquez=True, num_point=1000, ticks=24)
+    # FT_ = np.zeros(6)
+
+    plot_t = Thread(target=plotting_thread, args=[plot], name='plotting task')
```
Comment (lrozo) [inline examples/force_control/Force_control_example.py:227]: May you please explain the need to have threads here? I mean, I am curious about the motivation of using threads here for plotting and controlling the robot.

**Comment by TFLQW** [comment]
@lrozo I have updated the package that has solved the problem you have mentioned. 
When it's come to the thread pool issue, sorry I forgot why I have add these parts, maybe I guess I didn't figure out the plotting function in the original package, and for the drawing part, I referred to your original code about plotting, and created a plotting file which in thread method may have a good performance. I am not good at the realtime plotting in python, so maybe you could compact this part.

**Comment by TFLQW** [comment]
@lrozo Sorry for these detial issues, are the language problems solved, I forget whether somewhere is still not English

**Pull request merged**